Allow to create temporal objects from standard JS Date

lutovich opened the pull request:
This PR adds `#fromStandardDate()` functions to all temporal types except `Duration`. Such functions allow to create temporal objects from the provided standard JavaScript `Date`.

lutovich committed:
Allow to create temporal objects from standard JS Date

This commit adds `#fromStandardDate()` functions to all temporal types
except `Duration`. Such functions allow to create temporal objects
from the provided standard JavaScript `Date`.

ali-ince commented on src/v1/temporal-types.js:
```diff
@@ -168,6 +203,21 @@ export class Date {
+
+    return new Date(
+      standardDate.getFullYear(),
+      standardDate.getMonth(),
```
I think there's a problem about conversion of `month` value from/to standard date as javascript `Date` object stores months as `0` based values - with `0` = `January` and `11` = `December`. It seems to me that we do not consider this difference in conversions.

Maybe add value range checks on temporal type constructors?

ali-ince commented on src/v1/internal/temporal-util.js:
```diff
@@ -264,6 +265,27 @@ export function dateToIsoString(year, month, day) {
+}
+
+/**
+ * Get the total number of nanoseconds from the given standard JavaScript date.
```
It seems that the comment is outdated.

ali-ince commented on src/v1/temporal-types.js:
```diff
@@ -211,6 +261,26 @@ export class LocalDateTime {
+
+    return new LocalDateTime(
+      standardDate.getFullYear(),
+      standardDate.getMonth(),
```
Same here.

ali-ince commented on src/v1/temporal-types.js:
```diff
@@ -261,6 +331,27 @@ export class DateTime {
+
+    return new DateTime(
+      standardDate.getFullYear(),
+      standardDate.getMonth(),
```
Same here.

ali-ince commented on test/v1/temporal-types.test.js:
```diff
@@ -658,6 +659,147 @@ describe('temporal-types', () => {
     expect(() => new neo4j.types.DateTime(1, 2, 3, 4, 5, 6, 7, 8, 'UK')).toThrow();
   });
 
+  it('should convert standard Date to neo4j LocalTime', () => {
```
Maybe add some integration tests that incorporate standard date -> temporal type conversions?

lutovich committed:
Fix JSDoc

lutovich committed:
Fix handling of standard dates with zero month

Standard dates have zero-based month. Neo4j temporal types have
1-based month. Conversion from standard date with zero month was not
handled correctly and resulted in zero month in neo4j temporal types.

oskarhane reviewed (approved):
Good stuff, I have nothing to add!

ivan-kleshnin commented:
Any reason to not apply this to native JS dates automatically?

lutovich commented:
@ivan-kleshnin the problem is that driver can't know which Neo4j type should the native JS date be converted to. It can be converted to either `LocalDateTime` or `DateTime` but driver can't know if time zone information should be preserved or not. Technically, native JS dates can also get converted to `Date`, `LocalTime` or `Time` by using a subset of available getters. Only use `getFullYear()`, `getMonth()` and `getDate()` to construct Neo4j Date from a native date, ignore `getHours()`, etc.

That's why this PR adds explicit conversion methods so application code can tell driver which Neo4j temporal type is required.

Does this make sense to you?

ivan-kleshnin commented:
@lutovich yes, thanks. That what I thought, but as a Neo4j noob wanted to ensure :)
